### PR TITLE
chore(lib/trie): refactor `lib/trie/database.go`

### PR DESCRIPTION
**Pull request by qdm12**
## Changes

This is a simple 'linear' refactor in order to see diffs in a clearer way.

- Explicit variable names
- Less indentation (avoid type switches)
- More return early
- Add named returns
- Clarify comments
- Wrap errors

## Tests

```
go test github.com/ChainSafe/gossamer/lib/trie/... github.com/ChainSafe/gossamer/internal/trie/...
```

## Issues

- #2085

## Primary Reviewer

### CODE REVIEW

File: lib/trie/database.go
Comment (EclesioMeloJunior) [inline lib/trie/database.go:194]: Would be nice to log the index here
```suggestion
			return fmt.Errorf("cannot load child at index %d with hash 0x%x: %w", i, hash, err)
```

File: lib/trie/database.go
Comment (EclesioMeloJunior) [inline lib/trie/database.go:280]:
```suggestion
```

File: lib/trie/database.go
Comment (EclesioMeloJunior) [inline lib/trie/database.go:403]: @qdm12 could I create an issue to address this TODO?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:403]: Opened #2115 for it (pointing to this PR), it was quicker than writing an issue imo :smile_cat:

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go]: Yeah left it to try it out, I ran the node in production and never encountered a panic. I'll assume this never happen :wink:

**Review by noot — APPROVED** [review]

File: lib/trie/database.go
Comment (kishansagathiya) [inline lib/trie/database.go:294]: I am guessing this would have never happened, because go does not permit use of untyped nil `use of untyped nil`. parent has to be one of node or leaf.

**Review by kishansagathiya — APPROVED** [review]
:tada: 
I like that we are checking `!ok` instead of `ok`
Looks good, only small comments!

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:150]: rootHash is a `common.Hash` and its `String() string` method returns hexadecimal already :wink: 
Just re-using existing code hehe, although I agree with you it's not that obvious!

Most other places in this file uses raw `[]byte` where we do need `0x%x` however.

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:260]: Totally, changed from `0x%x` to `%s`

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:301]: Readability :wink: (aka smaller line length, more descriptive names for each subpart)

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:315]: Same as above, this is for readability. I used to nest calls to save on the number of lines but I now prefer to split in on more lines since I find it way more nice to read.

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:294]: It can happen if the parent is `nil`, for example with an empty trie the root can be `nil` (untyped).

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:203]: Changed to `PopulatedHashesSet`

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:72]: Done :+1:

**Comment by qdm12** [comment]
@kishansagathiya 

> I like that we are checking !ok instead of ok

Where do you mean this? :thinking:

**Comment by kishansagathiya** [comment]
@qdm12 

> like that we are checking !ok instead of ok

Sometimes, we check for the expected condition (true) and run when true and don't do anything when false.

I think you would have checked for the unexpected condition first and run some code on that unexpected condition.

**Comment by qdm12** [comment]
@kishansagathiya I tend to have the smaller code part inside the if block checking `ok` to reduce code indentation and improve  readability. Maybe do you have an example where this could be improved in this PR?

File: lib/trie/database.go
Comment (qdm12) [reply lib/trie/database.go:398]: I'll merge #2115 in development once this one is merged.

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**
